stl/inc/format

```diff
@@ -343,72 +337,395 @@ public:
+    case _Basic_format_arg_type::_Float_type:
+        return _Vis(_Arg._Float_state);
+    case _Basic_format_arg_type::_Double_type:
+        return _Vis(_Arg._Double_state);
+    case _Basic_format_arg_type::_Long_double_type:
+        return _Vis(_Arg._LDouble_state);
+    case _Basic_format_arg_type::_Pointer_type:
+        return _Vis(_Arg._Pointer_state);
+    case _Basic_format_arg_type::_CString_type:
+        return _Vis(_Arg._CString_state);
+    case _Basic_format_arg_type::_String_type:
+        return _Vis(_Arg._String_state);
+    case _Basic_format_arg_type::_Custom_type:
+        return _Vis(_Arg._Custom_state);
+    default:
+        _STL_ASSERT(false, "Invalid basic_format_arg type");
+        return _Vis(0);
+    }
+}
+
+template <class _Context, class _Ty>
+static constexpr auto _Get_format_arg_storage_type() noexcept {
+    using _CharType = typename _Context::char_type;
+    if constexpr (is_same_v<_Ty, monostate>) {
+        return monostate{};
+    } else if constexpr (is_same_v<_Ty, _CharType>) {
+        return _CharType{};
+    } else if constexpr (signed_integral<_Ty> && sizeof(_Ty) <= sizeof(int)) {
+        return int{};
+    } else if constexpr (unsigned_integral<_Ty> && sizeof(_Ty) <= sizeof(unsigned)) {
+        return unsigned{};
+    } else if constexpr (signed_integral<_Ty> && sizeof(_Ty) <= sizeof(long long)) {
+        return static_cast<long long>(42);
+    } else if constexpr (unsigned_integral<_Ty> && sizeof(_Ty) <= sizeof(unsigned long long)) {
+        return static_cast<unsigned long long>(42);
+    } else if constexpr (is_same_v<_Ty, bool>) {
+        return true;
+    } else if constexpr (is_same_v<_Ty, float>) {
+        return float{};
+    } else if constexpr (is_same_v<_Ty, double>) {
+        return double{};
+    } else if constexpr (is_same_v<_Ty, long double>) {
+        return static_cast<long double>(42);
+    } else if constexpr (is_same_v<_Ty, const void*>) {
+        return static_cast<const void*>(nullptr);
+    } else if constexpr (is_same_v<_Ty, const _CharType*>) {
+        return static_cast<const _CharType*>(nullptr);
+    } else if constexpr (is_same_v<_Ty, basic_string_view<_CharType>>) {
+        return basic_string_view<_CharType>{};
+    } else {
+        return basic_format_arg<_Context>::handle();
+    }
+}
+
+template <class _Context, class _Ty>
+static constexpr size_t _Get_format_arg_storage_size() noexcept {
+    return sizeof(_Get_format_arg_storage_type<_Context, _Ty>());
+}
+
+struct _Format_arg_store_packed_index {
+    // TRANSITION, Should be templated on number of arguments for even less storage
+    using _Index_type = size_t;
+
+    constexpr _Format_arg_store_packed_index() = default;
+    constexpr _Format_arg_store_packed_index(const size_t _Index)
+        : _Index(static_cast<_Index_type>(_Index)), _Type(_Basic_format_arg_type::_None) {}
+
+    _Index_type _Index : (sizeof(_Index_type) * 8 - 4);
+    _Basic_format_arg_type _Type : 4;

```

Comment:
note to self: this is better than
{
  data[]
  offset[]
  tag[]
}
because tag is half a byte and this way we don't pull in useless tags when we load the one we want